### ISSUE TITLE
Encrypt the master seed

### ISSUE THREAD

**Issue by plaprade** [haskoin/haskoin-wallet#5]
When a wallet is initialized, the seed is turned into a master extended private key (BIP32 format) and stored unencrypted. We need to encrypt that key.

**Comment by tphyahoo** [comment]
Haskoin wallet seems to expect online operation for most of its functionality.

Historically, encrypting wallets has never worked very well.  Hackers just switch to stealing the password when it is entered, rather than just stealing the data file.  Not much difference, once your system is pwned.

Wouldn't it make more sign to adopt a viewing / signing paradigm, and delegate signing operations to a trusted subsystem?

For low value wallets, the subsystem could be online, but there would also be the option of interfacing with hardware wallets or other more secure systems.

**Comment by xenog** [comment]
Hey @tphyahoo,

I have been thinking in a similar direction, delegating signing so that it happens outside of the `hw` process that is running on the “server side.” This would effectively make the server fully unprivileged, and allow for signing to be the cilent’s problem. It could be implemented in hardware, for example. I am open to design changes in this regard. Do you have something specific on your mind?

Greetings

**Comment by tphyahoo** [comment]
I'm not a heavy haskoin wallet user, just interested and watching the project. Don't even have it installed currently (though I have installed and played with it a while ago).  That being said... we have, from

https://github.com/haskoin/haskoin-wallet/blob/master/config/help

Transaction commands:
  .....
  import    acc tx                        Import a transaction (does not sign)
  sign      acc txid                      Sign one of your offline transactions

Offline tx signing commands:
  getoffline  acc txhash                  Get data to sign a tx offline
  signoffline acc tx coindata             Sign a tx with offline signing data

We'd want to start here to make it easy to work with trezor, ledger, and other bip32 hardware wallets.

Maybe import already does everything necessary.  Does this import and then broadcast signed transactions?

That being the case, you could close this issue and focus on having two modes of operation for hw: online and offline/signing. 

For offline/signing, the assumption is that this is either low value or happening on secure hardware.  

As a side note, it would be good to separate out functionality for signing (aka loading) and broadcasting transactions.  Electrum does this, and it be helpful / reassuring to review signed transactions before broadcast.

**Comment by tphyahoo** [comment]
More:

There could be various signing facilities

signoffline-haskoin-native acc tx coindata             Sign a tx with offline signing data <same as current>
signoffline-trezor <trezor compat>
signoffline-ledger
.... etc, maybe there are other options

Whenever a wallet is started in offline/signing mode, with signoffline-haskoin-native i.e. whenever a seed is loaded or generated, display a warning to user that seed is being stored unencrypted so this should only be done on secure hardware and preferably for low value wallets.

For other signing modes, no warning needed.

**Comment by xenog** [comment]
I was thinking of simplifying signing so that the `hw` process did not have to have any knowledge of private keys at all, and the client would have to perform signatures. Either the `hw` client would do it, or any client that connects to the ØMQ interface. Of course the command-line client could be extended to add support for (some) hardware wallets as well.

**Comment by tphyahoo** [comment]
Yes, if hw is an online server process, I think removing knowledge of private keys would be a good idea.

I agree, put it all in the client.

**Comment by tphyahoo** [comment]
Is there currently curently actually something that is an "hw client" or is something you would build now to satisfy these requirements?

seems to me the latter.

the hw process is essentially a server process when run in daemon mode. 

and (copying how electrum, libbitcoin sx, and many others do it) you would essentially have two clients.

An online client, which does things that a viewing wallet would do.  Create unsigned transactions, mainly.  

And a paranoid signing client, which only signs transactions and transmits them to the server for broadcast.  This wouldn't necessarily have to be offline, but might interface with a hardware wallet for highest security signatures.

Does this more or less fit with your vision?

**Comment by xenog** [comment]
Well, to simplify work, given that we do not have many developers, I would advocate to have the current client code in `hw` deal with private keys.

The `hw` executable contains both client and server code. Running `hw start` starts the server, while most other commands invoke client code, which requires a server to be available over a ØMQ socket.

Running the server on a remote machine while using the client to communicate with it is possible, although the ØMQ interface currently does not support encrypting or authenticating messages. In the current design it is generally expected that client and server are in the local machine communicating through a Unix socket whose access permissions are enforced by the operating system.

There is a haskoin-api package that creates a RESTful JSON API on top of the ØMQ socket, but this only works over TCP, making it inconvenient to use for IPC.

Due to the flexibility of ØMQ, perhaps we should standardize on it and add custom encryption and authorization layers. This would make it almost universal, but would leave browsers out, so still something like haskoin-api would be required to act as a bridge for HTTP-based clients.

This is all something that is yet to be discussed, I would love to get opinions and ideas.

**Comment by tphyahoo** [comment]
As a potential user and possible contributor, I'd like to see hw safe enough and convenient enough that it gets production use asap.  So, take the most direct path.

Looks like the encryption story for zmq is still being written.  There are 2 paths

vanilla zmq over ssl/tls
and zmq 4 / curveqmq, with something more custom and more secure/flexible

If I was going to choose one, I would lean mildly toward option 1, vanilla ssl/tls.

http://www.riskcompletefailure.com/2012/09/tls-and-zeromq.html

is a starting point for that path.

But I think we should also consider that hw is useful in production for a variety of scenarios, without wire encryption, as long as the keys are not on-server.

I say that hesitantly, and invite to be shot down if I'm overlooking something obvious.  But I think we should at least consider this.  

The case is, that a lot of information is public on the blockchain anyway.  Transactions and addresses.

So once you remove privkeys, what an attacker can gain over the wire is accounts and especially, account names. (Anything else?)

That's not good, but it's not an immediate theft scenario, like we have with privkeys on server.

So, I would break down current task into two steps

1 -- get privkeys off the server 
2 -- figure out a story for geting all messaging encrypted

It's possible that 1 is trivial with current functionality, and all it takes is updating the help file and readme.  The main thing that I see missing is something that lets you add a key ring using mpubkey rather than generating mprivkey on server.  In any case, it looks relatively straightforward in terms of functionality and doesn't really require more architecture

For task 2, this might be batted around a whille, since it's an architectural change, adding another layer of security and where does fit.  But meanwhile, we have a not 100% secure server, but for some applications could be accounts and account names info is public anyway, it's good enough.

**Comment by tphyahoo** [comment]
@xenog is the haskoin-api package public? 

I couldn't find it on github.com/haskoin

**Comment by tphyahoo** [comment]
further thoughts.

for concreteness, the machine running hw start is the server.

further, assume the server machine has a facility for loading pub only keyrings.

there is actually no need for this server to be account name aware.

all we need is a facility for selecting and caching pubseed relevant txouts, and some way for the clieny machine(s) to download the txouts.
further 
txouts selection and signing can then happen on the client boxrn, where there can be a vddy ariety of security policies, including air gapping or hardware wallets depending on the value of the privkey ring.

txouts are still communicated in the clear, but we can muddy the waters by using probabilistic bloom filters, as other txout servers (aka obelisk? electrum?) do, or maybe only talk about doing, I'm fuzzy on the details.

tldr, even if we're transacting jn the clear, we can avoid revealing account names, making this a much less interesting attack surface.

**Comment by tphyahoo** [comment]
in terms of the last proposal, main thing missing is a facility for client machine to request txouts associated with mpubkey, perhaps parametrized by block number, so we don't have to go back to beginning of time.  

maybe this could be winnowed further by requesting particular bip32 branches.
or maybe we don't want to reveal mpubkey, only addresses. a little more work, but should be ok too.

so
-- hw server becomes essentially mpubkey (or address) to txout caching service, always on and watching for txs.
-- hw client does txout selection, signing, and is responsible for protecting privkeys, and possibly mpubkeys if we're paranoid about leaking those.

kind of like electrum, but easier to hack on.

thoughts?

**Comment by tphyahoo** [comment]
some adjustments may be in order to what I wrote previously.

is hw already using bip37 bloom filters and get relevant transaction history and ignore the bulk? (I guess it is, can someone confirm?)

if so much of what I said about caching, and privacy, can be disregarded.

further, to provide a fully encrypted comms channel to early production users, I think hw client commands over ssh is sufficient, and armoring zmq can be left as a separate task. this may require some additional hw commands for supporting common use cases but is probably easier than protecting zmq.

(as a side note, I couldn't find any documentation for trying out the zmq facility. pointers?)

if all of the above is true, could be that all that is required to use hw server safely in production for basic use cases is a facility for setting appropriate getdata bloom filters with mpubkey.

that doesn't sound that hard.

**Comment by xenog** [comment]
Hey @tphyahoo,

First, I answer your questions:

- BIP-37 bloom filters are being used
- ØMQ is using JSON as serialization format
- The ØMQ server handler is at `Haskoin/Wallet/Server/Handler.hs`

Then, to see if I understood correctly, I will paraphrase your proposal:

The client:
- selects coins,
- signs,
- handles private keys

The server:
- tracks transactions pertaining to extended public keys provided by clients
- provides unspent outputs to client upon request
- broadcast transactions on behalf of client

Since the responsibility for signing and preparing transactions never leaves the client process, there is no risk of theft if communication with server happens over a non-encrypted channel. The server does little more than index and relay publicly-available information between the blockchain/mempool and the client.

For increased privacy, and to validate the server identity, TLS may be used.

Although the server has knowledge of extended public keys provided from the client, nobody outside the server should be able to obtain this information, so privacy leaks are contained. It is still better to track extended public keys and not individual addresses to improve bandwidth utilization when the client communicates with the server.

I like it.

**Comment by tphyahoo** [comment]
@xenog glad you like! I think it's a good compromise, and a pragmatic approach.

loose ends

-- can you point me to haskoin-api package mentioned above?

-- there is a non-obvious security  issue with mpubkeys that users should perhaps be warned about if they are placing their mpubkeys in a shared or low security environment.

"One weakness that may not be immediately obvious, is that knowledge of a parent extended public key plus any non-hardened private key descending from it is equivalent to knowing the parent extended private key (and thus every private and public key descending from it). "

https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#Security

So basically leaking any private key is equivalent to leaking the entire wallet, modulo hardening.

I think it's an acceptable risk, but users need to be aware of it.

Another point of view is that mpubkeys are too sensitive to keep in a shared environment or potentially insecure environment, and only addresses should be shared with the transaction caching server.

I suppose for maximum flexibility, either approach would be allowed.  In this case I would advocate to first implement the mpubkey viewing wallet version first, and then the addresses only caching version as a later enhancement, so as to get to something usable quicker.



@xenog glad you like! I think it's a good compromise, and a pragmatic approach.

loose ends

-- can you point me to haskoin-api package mentioned above?

-- there is a non-obvious security  issue with mpubkeys that users should perhaps be warned about if they are placing their mpubkeys in a shared or low security environment.

"One weakness that may not be immediately obvious, is that knowledge of a parent extended public key plus any non-hardened private key descending from it is equivalent to knowing the parent extended private key (and thus every private and public key descending from it). "

https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki#Security

So basically leaking any private key is equivalent to leaking the entire wallet, modulo hardening.

I think it's an acceptable risk, but users need to be aware of it.

Another point of view is that mpubkeys are too sensitive to keep in a shared environment or potentially insecure environment, and only addresses should be shared with the transaction caching server.

I suppose for maximum flexibility, either approach would be allowed.  In this case I would advocate to first implement the mpubkey viewing wallet version first, and then the addresses only caching version as a later enhancement, so as to get to something usable quicker.

**Comment by xenog** [comment]
We haven’t published the haskoin-api package yet. I’ll get it published when we align it with the latest codebase, if it makes sense to do so. That package has been archived for almost a year. To be honest, I think we should not use it, but focus on getting these changes directly into the haskoin-wallet package.

I know that revealing an extended public key and the private key of one of the addresse exposes all private keys in the tree. It would make more sense to only share the extended public key with the server which will track transactions, and communicate in an encrypted manner, authenticating the server through TLS certificates.

Our client software should never expose individual private keys anyway. It should not be allowed. If somebody wants to do it, then they’ll have to derive the keys themselves. It should be hard to shoot yourself in the foot.

I think we should go with supporting extended keys for maximum flexibility and bandwidth utilization efficiency. If somebody wants to use a fully-indexed blockchain service for individual addresses, there are many services out there that can be used for that, such as Electrum Stratum servers, BitPay Insight servers, and Blockchain.info, among many many others.

**Comment by tphyahoo** [comment]
Sounds good.

Close this ticket and start opening other tickets, in line with current plan?

**Comment by xenog** [comment]
Sounds good to me.
